File: python/code_challenges/sorting_algorithms/merge_sort.py

def merge_sort(start_list):
  # if statement to break every sublist down to 1
  if len(start_list) > 1: 
    # find middle point of lists
    middle = len(start_list) // 2
    # create left sublist with everything before the middle
    left_list = start_list[:middle]
    # create right sublist with everything middle and after
    right_list = start_list[middle:]
    # recursively run function on right sublist
    merge_sort(right_list)
    # recursively run function on left sublist
    merge_sort(left_list)
    # create variables to hold indexes of left,right, and starting lists respectively
    i = 0 # left
    j = 0 # right
    k = 0 # starting list
    # while loop to that goes until list is merged
    while i < len(left_list) and j < len(right_list):
      # check if current index in left sublist is less than current index in right sublist
      if left_list[i] < right_list[j]:
        # if true, move element to current index in starting list
        start_list[k] = left_list[i]
        # move to next index in left sublist
        i += 1
        # move to next index in starting sublist
        k += 1
      else:
        # if not, move element from right sublist to starting list
        start_list[k] = right_list[j]
        # move to next index in left sublist
        j += 1
        # move to next index in starting sublist
        k += 1
    # add leftovers to end of list
    while i < len(left_list):
      start_list[k] = left_list[i]
      i += 1
      k += 1
    while j < len(right_list):
      start_list[k] = right_list[j]
      j += 1
      k += 1

File: javascript/code-challenges/merge-sort/mergeSort.js
'use strict';

const mergeSort = (startArray) => {
  // if statement to break every subarray down to 1
  if (startArray.length > 1) {
    // find middle point of array
    let middle = Math.ceil(startArray.length / 2);
    //create left subarray with everything before the middle
    let leftArray = startArray.slice(0, middle);
    //create left subarray with everything before the middle
    let rightArray = startArray.slice(middle, startArray.length);
    // recursively run function on right and leftsubarrays
    mergeSort(rightArray);
    // eslint-disable-next-line no-unused-vars
    mergeSort(leftArray);
    //create variables to hold indexes of left,right, and starting arrays respectively
    let i = 0;//left
    let j = 0;//right
    let k = 0;//startArray
    // while loop to that goes until array is merged
    while ((i < leftArray.length) && (j < rightArray.length)){
      // check if current index in left subarray is less than current index in right subarray
      if (leftArray[i] < rightArray[j]){
        // if true, move element to current index in starting array
        startArray[k] = leftArray[i];
        // move to next index in left subarray
        i += 1;
        // move to next index in starting subarray
        k += 1;
      } else {
        // if not, move element from right subarray to starting subarray
        startArray[k] = rightArray[j];
        // move to next index in left subarray
        j += 1;
        // move to next index in starting subarray
        k += 1;
      }
    }
    // add leftovers to end of array
    while (i < leftArray.length) {
      startArray[k] = leftArray[i];
      i += 1;
      k += 1;
    }
    while (j < rightArray.length) {
      startArray[k] = rightArray[j];
      j += 1;
      k += 1;
    }
  }
};

module.exports = mergeSort;
